imagescan/controller.go

```diff
@@ -339,59 +351,62 @@ func (s *Controller) concurrentScansNumber() int {
 	return int(s.cfg.MaxConcurrentScans)
 }
 
-func (s *Controller) sendImagesResourcesChanges(ctx context.Context) {
-	images := s.delta.getImages()
+func (s *Controller) updateImageStatuses(ctx context.Context, images []*image, fullSnapshot bool) error {
+	if len(images) == 0 {
+		return nil
+	}
+	now := s.timeGetter()
 	var imagesChanges []castai.Image
 	for _, img := range images {
-		if img.ownerChanges.empty() {
-			continue
+		changedResourceIds := lo.Uniq(img.ownerChanges.addedIDS)
+		if fullSnapshot {
+			changedResourceIds = lo.Keys(img.owners)
+		}
+		var updatedStatus castai.ImageScanStatus
+		if IsImagePending(img, now) {
+			updatedStatus = castai.ImageScanStatusPending
 		}
 		imagesChanges = append(imagesChanges, castai.Image{
 			ID:           img.id,
 			Architecture: img.architecture,
 			ResourcesChange: castai.ResourcesChange{
-				ResourceIDs: lo.Uniq(img.ownerChanges.addedIDS),
+				ResourceIDs: changedResourceIds,
 			},
+			Status: updatedStatus,
 		})
 	}
-	if len(imagesChanges) == 0 {
-		return
-	}
 
 	s.log.Info("sending images resources changes")
-	report := &castai.ImagesResourcesChange{
-		Images: imagesChanges,
-	}
-	if err := s.client.SendImagesResourcesChange(ctx, report); err != nil {
-		s.log.Errorf("sending images resources changes: %v", err)
-		return
+	report := &castai.UpdateImagesStatusRequest{
+		FullSnapshot: fullSnapshot,
+		Images:       imagesChanges,
 	}
 
-	// Clear changes state.
-	for _, img := range images {
-		if img.ownerChanges.empty() {
-			continue
-		}
-		img.ownerChanges.clear()
-	}
+	return s.client.UpdateImageStatus(ctx, report)
 }
 
-func (s *Controller) sendFullSnapshotImageResources(ctx context.Context) error {
-	s.log.Info("sending initial full images resources changes")
-	images := s.delta.getImages()
-	report := &castai.ImagesResourcesChange{
-		FullSnapshot: true,
+func (s *Controller) updateImageStatusAsFailed(ctx context.Context, image *image, scanJobError error) error {
+	if image == nil {
```

Comment:
How can this be nil? This check can be removed as image passed here can't be nil.

Reply:
extra check won't harm in my opinion

Reply:
The issue that it's confusing. Now it looks like image is optional argument which if not provided will result in no op. But this can actually hide bugs as image should always be passed here.

To make it correct you will need to return error like `errors.New("image is missing")`. But it looks redundant it this case to return error.